File: Difficulty: Hard/Aggressive Cows/aggressive-cows.cpp
//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution {
public:

    int solve(int n, int k, vector<int> &stalls) {
        // Write your code here
        // 05.07.24 
        sort(stalls.begin(), stalls.end());
        
        int lo=1, hi=(stalls.back()-stalls[0])/(k-1);
        
        while(lo <= hi){
            int mid = lo + (hi-lo)/2;
            
            if(placementPossible(k, stalls, n, mid))
                lo = mid+1;
            else
                hi = mid-1;
        }
        
        return hi;
    }
    
    bool placementPossible(int k, vector<int>& stalls, int n, int dist){
        int count=1, lastStall=stalls[0];
        
        for(int i=1; i<n; i++){
            if(stalls[i] - lastStall >= dist){
                lastStall = stalls[i];
                count++;
            }
            
            if(count == k)
                return true;
        }
        
        return false;
    }
};

//{ Driver Code Starts.

int main() {
    int t = 1;
    cin >> t;

    // freopen ("output_gfg.txt", "w", stdout);

    while (t--) {
        // Input

        int n, k;
        cin >> n >> k;

        vector<int> stalls(n);
        for (int i = 0; i < n; i++) {
            cin >> stalls[i];
        }
        // char ch;
        // cin >> ch;

        Solution obj;
        cout << obj.solve(n, k, stalls) << endl;

        // cout << "~\n";
    }
    // fclose(stdout);

    return 0;
}
// } Driver Code Ends

File: Difficulty: Medium/Aggressive Cows/aggressive-cows.py
#User function Template for python3
class Solution:
    def aggressiveCows(self, stalls, k):
        # 17.12.24 poTD
        
        def canPlaceCows(stalls, dist, k):
            n = len(stalls)
            count = 1
            prevPos = stalls[0]
            
            for i in range(1,n):
                if stalls[i] - prevPos >= dist:
                    count += 1
                    prevPos = stalls[i]
                
                if count >= k:
                    return True
                    
            return False
            
        
        stalls.sort()
        ans, start, end = 0, 1, stalls[-1] - stalls[0]
        while start <= end:
            mid = (start + end)//2
            
            if canPlaceCows(stalls, mid, k):
                start = mid + 1
                ans = mid
            else:
                end = mid - 1
                
        return ans



#{ 
 # Driver Code Starts
#Initial Template for Python 3
import bisect
#Main
if __name__ == '__main__':
    t = int(input())
    while t:
        t -= 1
        A = [int(x) for x in input().strip().split()]
        nd = [int(x) for x in input().strip().split()]
        D = nd[0]
        ob = Solution()
        ans = ob.aggressiveCows(A, D)
        print(ans)
        print("~")
# } Driver Code Ends